### PR TITLE
feat(api-reference): add typedoc plugins for api reference

### PR DESCRIPTION
**Pull request by AllanZhengYP**
*Description of changes:*
This change adds 2 typedoc plugins:
1. `client-documentation-generator`: It is used when generating each client level api reference
Screen Shot 2020-11-17 at 11 45 19 PM

2. Private ` core-packages-documentation-generator`: It generates the landing page for the SDK. It landing page includes links to docs of all the dependency packages in `packages`, and links to all the docs for each client:
Screen Shot 2020-11-17 at 11 47 09 PM
Screen Shot 2020-11-17 at 11 47 27 PM

So the overall API reference is generated in 2 batches: 
1. Build each client's doc by `yarn build-documentation --out docs`. It will generate docs for each client under `docs` folder
2. Build packages docs and link the clients module to the landing page. In the project root: `yarn build-documentation --clientDocs clients/{{CLIENT}}/docs`

In each step, users can specify the output path with `out` argument.

TODO: 
1. Regenerate clients with correct Typedoc options

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -25,6 +25,19 @@
...
-  "exclude": ["node_modules/"]
+  "include": ["packages/", "lib/"],
+  "exclude": ["node_modules/"],
+  "typedocOptions": {
+    "readme": "README.md",
+    "mode": "modules",
+    "out": "docs",
+    "exclude": ["**/node_modules/**", "**/*.spec.ts"],
+    "excludeExternals": true,
+    "name": "AWS SDK v3",
```
Comment (trivikr) [inline tsconfig.json:36]: add JavaScript?
```suggestion
    "name": "AWS SDK for JavaScript v3",
```

File: packages/node-http-handler/typedoc.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+const { __esModule } = require("typedoc-plugin-lerna-packages/dist/plugin");
```
Comment (trivikr) [inline packages/node-http-handler/typedoc.js:1]: Why is `typedoc.js` required in node-http-handler?

File: packages/node-http-handler/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (trivikr) [inline packages/node-http-handler/tsconfig.json:1]: Why is `tsconfig.json` required in node-http-handler?

File: packages/core-packages-documentation-generator/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { ParameterType, PluginHost } from "typedoc/dist/lib/utils";
+
+import { SdkIndexLinkClientPlugin } from "./sdk-index-link-client";
+
+/**
+ *
```
Comment (trivikr) [inline packages/core-packages-documentation-generator/src/index.ts:6]: nit: line can be removed.
```suggestion
```

File: packages/core-packages-documentation-generator/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ * @param pluginHost An instance of PluginHost.
+ */
+module.exports = function load(pluginHost: PluginHost) {
+  const application = pluginHost.owner;
+  // Core packages doc generator plugins
```
Comment (trivikr) [inline packages/core-packages-documentation-generator/src/index.ts:11]: nit: period and a newline.
```suggestion

  // Core packages doc generator plugins.
```

File: packages/core-packages-documentation-generator/src/sdk-index-link-client.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  initialize() {
+    this.listenTo(this.owner, {
+      [PageEvent.BEGIN]: this.onPageBegin,
+    });
+  }
+
+  onPageBegin(page: PageEvent) {
+    const out = isAbsolute(this.out) ? this.out : resolve(PROJECT_ROOT, this.out);
+    const clientDocs = isAbsolute(this.clientDocs) ? this.clientDocs : resolve(PROJECT_ROOT, this.clientDocs);
+    // Get relative path from core packages doc to clients' doc
```
Comment (trivikr) [inline packages/core-packages-documentation-generator/src/sdk-index-link-client.ts:28]: nit: period and a newline.
```suggestion

    // Get relative path from core packages doc to clients' doc.
```

File: packages/core-packages-documentation-generator/src/sdk-index-link-client.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        .forEach((child) => {
+          // "clients/client-s3" => "client-s3"
+          const clientName = child.sources[0].fileName.split(sep)[1];
+          const clientDocDir = clientDocsPattern.replace(/{{CLIENT}}/g, clientName);
+          child.url = join(clientDocDir, "index.html");
+          //@ts-ignore attach temporary flag
+          child._skipRendering = true;
+        });
+    }
+    // Skip rendering empty landing page for each client
```
Comment (trivikr) [inline packages/core-packages-documentation-generator/src/sdk-index-link-client.ts:43]: nit: period and newline.
```suggestion

    // Skip rendering empty landing page for each client.
```

File: packages/core-packages-documentation-generator/src/sdk-index-link-client.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    if (page.model === page.project) {
+      // Entry point index.html and global.html page.
+      page.project.children
+        .filter((child) => child.sources[0].fileName.startsWith(`clients${sep}`))
+        .forEach((child) => {
+          // "clients/client-s3" => "client-s3"
+          const clientName = child.sources[0].fileName.split(sep)[1];
+          const clientDocDir = clientDocsPattern.replace(/{{CLIENT}}/g, clientName);
+          child.url = join(clientDocDir, "index.html");
+          //@ts-ignore attach temporary flag
```
Comment (trivikr) [inline packages/core-packages-documentation-generator/src/sdk-index-link-client.ts:39]: nit: space at the beginning of comment and period at the end.
```suggestion
          // @ts-ignore attach temporary flag.
```

File: packages/client-documentation-generator/src/sdk-client-rename-project.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { readFileSync } from "fs";
+import { Component, RendererComponent } from "typedoc/dist/lib/output/components";
+import { RendererEvent } from "typedoc/dist/lib/output/events";
+
+/**
+ * Correct the package name in the navigator
```
Comment (trivikr) [inline packages/client-documentation-generator/src/sdk-client-rename-project.ts:6]: nit: period.
```suggestion
 * Correct the package name in the navigator.
```

File: packages/client-documentation-generator/src/sdk-client-toc-plugin.ts [hunk 10]
```diff
@@ -4,11 +4,14 @@ import { Component, RendererComponent } from "typedoc/dist/lib/output/components
 import { PageEvent } from "typedoc/dist/lib/output/events";
 import { NavigationItem } from "typedoc/dist/lib/output/models/NavigationItem";
 
+/**
+ * Group the ToC for easier observability
```
Comment (trivikr) [inline packages/client-documentation-generator/src/sdk-client-toc-plugin.ts:8]: nit: period.
```suggestion
 * Group the ToC for easier observability.
```

File: packages/client-documentation-generator/src/sdk-client-toc-plugin.ts [hunk 11]
```diff
@@ -40,36 +43,43 @@ export class SdkClientTocPlugin extends RendererComponent {
...
 
-  private isCommand({ implementedTypes = [] }: DeclarationReflection): boolean {
+  private isClient(model: DeclarationReflection): boolean {
+    const { extendedTypes = [] } = model;
     return (
-      implementedTypes.length === 1 &&
-      implementedTypes[0].type === "reference" &&
-      (implementedTypes[0] as ReferenceType).name === "Command"
+      model.kindOf(ReflectionKind.Class) &&
+      model.getFullName() !== "Client" && // Exclude the Smithy Client class
```
Comment (trivikr) [inline packages/client-documentation-generator/src/sdk-client-toc-plugin.ts:58]:
```suggestion
      model.getFullName() !== "Client" && // Exclude the Smithy Client class.
```

File: packages/client-documentation-generator/src/sdk-client-toc-plugin.ts [hunk 12]
```diff
@@ -40,36 +43,43 @@ export class SdkClientTocPlugin extends RendererComponent {
...
 
-  private isException(model: DeclarationReflection): boolean {
-    const extendedTypes = model.extendedTypes || [];
+  private isCommand(model: DeclarationReflection): boolean {
     return (
-      extendedTypes.length === 1 &&
-      extendedTypes[0].type === "reference" &&
-      (extendedTypes[0] as ReferenceType).name === "ServiceException"
+      model.kindOf(ReflectionKind.Class) &&
+      model.getFullName() !== "Command" && // Exclude the Smithy Command class
```
Comment (trivikr) [inline packages/client-documentation-generator/src/sdk-client-toc-plugin.ts:68]:
```suggestion
      model.getFullName() !== "Command" && // Exclude the Smithy Command class.
```

File: packages/client-documentation-generator/src/sdk-client-toc-plugin.ts [hunk 13]
```diff
@@ -92,57 +102,33 @@ export class SdkClientTocPlugin extends RendererComponent {
...
           const item = NavigationItem.create(child, parent, true);
-          if (trail.indexOf(child) !== -1) {
+          if (trail.includes(child)) {
             item.isInPath = true;
             item.isCurrent = trail[trail.length - 1] === child;
             this.buildToc(child, trail, item);
           }
         }
       });
+      // Group commands and input/output interface of each command
```
Comment (trivikr) [inline packages/client-documentation-generator/src/sdk-client-toc-plugin.ts:130]:
```suggestion
      // Group commands and input/output interface of each command.
```

File: packages/client-documentation-generator/src/sdk-client-comment-update.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  private onDeclaration(context: Context, reflection: Reflection, node?: ts.Node) {
+    if (!node) return;
+    const rawComment = getRawComment(node);
+    if (!rawComment) return;
+    const comment = parseComment(this.cleanEmptyCommentLines(rawComment));
+    reflection.comment = comment;
+  }
+
+  /**
+   * Update documentation block to exclude empty lines
```
Comment (trivikr) [inline packages/client-documentation-generator/src/sdk-client-comment-update.ts:28]:
```suggestion
   * Update documentation block to exclude empty lines.
```

**Review by trivikr — COMMENTED** [review]
nit: update title to **feat(docs)**?

File: packages/node-http-handler/tsconfig.json (hunk 3 above)
Comment (AllanZhengYP) [reply packages/node-http-handler/tsconfig.json:1]: good catch. Will remove

**Review by trivikr — APPROVED** [review]

**Pull request merged**